pkg/autoscaler/stats_scraper.go

```diff
@@ -52,50 +59,114 @@ var cacheDisabledClient = &http.Client{
+	}
+	return newServiceScraperWithClient(metric, dynamicConfig, endpointsInformer, cacheDisabledClient)
 }
 
-func newServiceScraperWithClient(metric *Metric, logger *zap.SugaredLogger, httpClient *http.Client) (*ServiceScraper, error) {
+func newServiceScraperWithClient(
+	metric *Metric,
+	dynamicConfig *DynamicConfig,
+	endpointsInformer corev1informers.EndpointsInformer,
+	httpClient *http.Client) (*ServiceScraper, error) {
 	revName := metric.Labels[serving.RevisionLabelKey]
 	if revName == "" {
 		return nil, fmt.Errorf("no Revision label found for Metric %s", metric.Name)
 	}
+	if endpointsInformer == nil {
+		return nil, errors.New("Empty interface of EndpointsInformer")
+	}
 
 	serviceName := reconciler.GetServingK8SServiceNameForObj(revName)
 	return &ServiceScraper{
-		httpClient: httpClient,
-		url:        fmt.Sprintf("http://%s.%s:%d/metrics", serviceName, metric.Namespace, v1alpha1.RequestQueueMetricsPort),
-		metricKey:  NewMetricKey(metric.Namespace, metric.Name),
-		logger:     logger,
+		httpClient:      httpClient,
+		endpointsLister: endpointsInformer.Lister(),
+		url:             fmt.Sprintf("http://%s.%s:%d/metrics", serviceName, metric.Namespace, v1alpha1.RequestQueueMetricsPort),
+		metricKey:       NewMetricKey(metric.Namespace, metric.Name),
+		namespace:       metric.Namespace,
+		revisionService: reconciler.GetServingK8SServiceNameForObj(revName),
+		cachedReadyPods: readyPods{},
 	}, nil
 }
 
 // Scrape call the destination service then send it
 // to the given stats chanel
-func (s *ServiceScraper) Scrape(statsCh chan<- *StatMessage) {
+func (s *ServiceScraper) Scrape(ctx context.Context, statsCh chan<- *StatMessage) {
+	logger := logging.FromContext(ctx)
+
+	readyPods, err := s.readyPods(time.Now())
+	if err != nil {
+		logger.Errorw("Failed to get Endpoints via K8S Lister", zap.Error(err))
+		return
+	}
+
+	if readyPods == 0 {
+		logger.Debug("No ready pods found, not to scrape.")
+		return
+	}
+
 	stat, err := s.scrapeViaURL()
 	if err != nil {
-		s.logger.Errorw("Failed to get metrics", zap.Error(err))
+		logger.Errorw("Failed to get metrics", zap.Error(err))
 		return
 	}
 
+	// Assume traffics are route to pods evenly. A particular pod can stand for
+	// other pods, i.e. other pods have similar concurrency.
+	stat.AverageRevConcurrency = stat.AverageConcurrentRequests * float64(readyPods)
+
 	s.sendStatMessage(*stat, statsCh)
 }
 
+// readyPods returns the ready IP count in the K8S Endpoints object for a Revision
+// via K8S Informer. This is same as ready Pod count.
+func (s *ServiceScraper) readyPods(now time.Time) (int, error) {
+	if s.cachedReadyPods.time.Add(cacheTimeout).After(now) {
+		return s.cachedReadyPods.count, nil
+	}
+
+	readyPods := 0
+	endpoints, err := s.endpointsLister.Endpoints(s.namespace).Get(s.revisionService)
+	if apierrors.IsNotFound(err) {
+		// Treat not found as zero endpoints, it either hasn't been created
+		// or it has been torn down.
```

Comment:
so this no longer will return "1", as in other places for division purposes. Is that intended?

Reply:
Yes. Here we do not scrape if there are no pods. When calculate revision concurrency, we report 1 if there are no pods for multiplication. I found we only use `readyPods` for scale up limitation. I changed another place to return 0 as well and use 1 as minimum where it is used.